adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,508 @@
+        tlsio_instance->on_io_error_context = on_io_error_context;
+
+        if ((tlsio_instance->state == TLSIO_APPLEIOS_STATE_ERROR) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPENING) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPEN) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_CLOSING))
+        {
+            /* Codes_SRS_TLSIO_APPLEIOS_32_035: [ If the tlsio state is TLSIO_APPLEIOS_STATE_ERROR, TLSIO_APPLEIOS_STATE_OPENING, TLSIO_APPLEIOS_STATE_OPEN, or TLSIO_APPLEIOS_STATE_CLOSING, the tlsio_appleios_open shall set the tlsio state as TLSIO_APPLEIOS_STATE_ERROR, and return _LINE_. ]*/
+            LogError("Try to open a connection with an already opened TLS.");
+            tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+            result = __FAILURE__;
+        }
+        else
+        {
+		    CFStreamStatus status;
+			
+			status = (tlsio_instance->sockRead == NULL)? kCFStreamStatusClosed : CFReadStreamGetStatus(tlsio_instance->sockRead);
+			
+			if (status != kCFStreamStatusClosed)
+            {
+                /* Codes_SRS_TLSIO_APPLEIOS_32_037: [ If the ssl client is connected, the tlsio_appleios_open shall change the state to TLSIO_APPLEIOS_STATE_ERROR, log the error, and return _LINE_. ]*/
+                LogError("SSL stream is not closed.");
+                tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+                result = __FAILURE__;
+            }
+			else
+			{
+	            CFStreamCreatePairWithSocketToHost(NULL, tlsio_instance->remote_host, tlsio_instance->port, &tlsio_instance->sockRead, &tlsio_instance->sockWrite);
+				
+				if (tlsio_instance->sockRead == NULL || tlsio_instance->sockWrite == NULL)
+				{
+					LogError("Unable to create socket streams.")
+					tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+					result = __FAILURE__;
+				}
+				else if (!CFReadStreamSetProperty(tlsio_instance->sockRead, kCFStreamPropertySSLSettings, kCFStreamSocketSecurityLevelTLSv1))
+				{
+					LogError("Unable to start SSL on socket.");
+					tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+					result = __FAILURE__;
+				}
+	            else if (!CFReadStreamOpen(tlsio_instance->sockRead) || !CFWriteStreamOpen(tlsio_instance->sockWrite))
+				{
+					LogError("Failure: connect failure");
+					CFReadStreamClose(tlsio_instance->sockRead);
+					CFWriteStreamClose(tlsio_instance->sockWrite);
+					tlsio_instance->sockRead = NULL;
+					tlsio_instance->sockWrite = NULL;
+					result = __FAILURE__;
+				}
+				else
+				{
+					tlsio_instance->state = TLSIO_APPLEIOS_STATE_OPENING;
+				}
+			}
+		}
+    }
+
+    if (result == 0)
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_041: [ If the tlsio_appleios_open get success opening the tls connection, it shall call the tlsio_appleios_dowork. ]*/
+        tlsio_appleios_dowork(tlsio_handle);
```

Comment:
>tlsio_appleios_dowork [](start = 8, length = 21)

This should not be done. The user is in charge of scheduling work, not the TLS IO.
Whenever we place any dowork somewhere in the code, we end upregretting it (it pops up somehow as a bug later. Case in point httpapi_compact, umqtt close, etc.

So, please remove it and its requirement Codes_SRS_TLSIO_APPLEIOS_32_041